FEXHeaderUtils/FEXHeaderUtils/Filesystem.h

```diff
@@ -14,4 +22,270 @@ namespace FHU::Filesystem {
+      const auto Size = End - Begin;
+      ExpectedStringSize += Size;
+
+      // If Size is zero then only insert if Parts is empty.
+      // Ensures that we don't remove an initial `/` by itself
+      if (Parts.empty() || Size != 0) {
+        Parts.emplace_back(std::string_view(Begin, End));
+      }
+
+      if (Size == 0) {
+        // If the view is empty, skip over the separator.
+        FoundSeperator += 1;
+      }
+
+      CurrentOffset = FoundSeperator;
+    } while (CurrentOffset != PathSize);
+
+    size_t CurrentIterDistance{};
+    for (auto iter = Parts.begin(); iter != Parts.end();) {
+      if (*iter == ".") {
+        // Erase '.' directory parts if not at root.
+        if (CurrentIterDistance > 0) {
+          // Erasing this iterator, don't increase iter distances
+          iter = Parts.erase(iter);
+          --ExpectedStringSize;
+          continue;
+        }
+      }
+
+      if (*iter == "..") {
+        auto Distance = std::distance(Parts.begin(), iter);
+        if (Distance > 0) {
+          // If not at root then remove both this iterator and the previous one.
+          // ONLY if the previous iterator is also not ".."
+          //
+          // If the previous iterator is '.' then /only/ erase the previous iterator.
+          auto PreviousIter = iter;
+          --PreviousIter;
+
+          if (*PreviousIter == ".") {
+            // Erasing the previous iterator, iterator distance has subtracted by one
+            --CurrentIterDistance;
+            ExpectedStringSize -= PreviousIter->size();
+            Parts.erase(PreviousIter);
+          }
+          else if (*PreviousIter != "..") {
+            // Erasing the previous iterator, iterator distance has subtracted by one
+            // Also erasing current iterator, which means iterator distance also doesn't increase by one.
+            --CurrentIterDistance;
+            ExpectedStringSize -= PreviousIter->size() + 2;
+            Parts.erase(PreviousIter);
+            iter = Parts.erase(iter);
+            continue;
+          }
+        }
+      }
+
+      // Interator distance increased by one.
+      ++CurrentIterDistance;
+      ++iter;
+    }
+
+    fextl::string OutputPath{};
+    OutputPath.reserve(ExpectedStringSize + Parts.size());
+
+    auto iter = Parts.begin();
+    for (size_t i = 0; i < Parts.size(); ++i, ++iter) {
+      auto &Part = *iter;
+      OutputPath += Part;
+
+      const bool IsFinal = (i + 1) == Parts.size();
+      // If the final element is ellipses then don't apply a separator.
+      // Otherwise if it is anything else, apply a separator.
+      const bool IsEllipses = Part == "." || Part == "..";
+
+      const bool NeedsSeparator =
+        !IsFinal || // Needs a separator if this isn't the final part.
+        (IsFinal && !IsEllipses); // Needs a separator if the final part doesn't end in `.` or `..`
+
+      if (NeedsSeparator) {
+        OutputPath += '/';
+      }
+    }
```

Comment:
Looks like this could be simplified a lot by using `fextl::format("{}", fmt::join(Parts, '/'))` combined with a special-case for `IsEllipses`.

Reply:
That wouldn't really be feasible with how the logic works.

Reply:
What speaks against this?
```cpp
auto OutputPath = fextl::format("{}", fmt::join(Parts, '/'))
if (!Parts.empty && Parts.back() != "." && Parts.back() != "..") {
  // Add a separator unless the final element is ellipses.
  OutputPath += '/';
}
```

Reply:
Switched over to it. For me it's harder to reason about but oh well.